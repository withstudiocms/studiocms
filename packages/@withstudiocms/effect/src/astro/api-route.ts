import type { APIContext, APIRoute } from 'astro';
import { runEffect } from '../index.js';
import type {
	EffectAPIRouteHandler,
	EffectRouteOptions,
	ExtractDefinedRoutes,
	RouteHandlerConfig,
	RouteHandlers,
} from './types.js';
import { applyCors, validateRequest } from './utils/api-route.js';

/**
 * Defines an API route handler for Astro using an effect-based approach.
 *
 * @param context - The API context provided by Astro for the current request.
 * @returns A function that takes an `EffectAPIRouteHandler`, executes it with the provided context,
 *          and returns a `Promise<Response>` representing the HTTP response.
 *
 * @example
 * ```typescript
 * export const GET = defineAPIRoute(context)(async (ctx) => {
 *   // Your effectful logic here
 *   return new Response("Hello, world!");
 * });
 * ```
 *
 * @deprecated Use `createEffectRoute` instead for better clarity and consistency.
 */
export const defineAPIRoute =
	(context: APIContext) =>
	async (fn: EffectAPIRouteHandler): Promise<Response> =>
		await runEffect(fn(context));

/**
 * Creates an Astro API route handler that executes the provided effectful handler function.
 *
 * @param fn - The effectful API route handler function to be executed. It should accept an `APIContext` and return an effect.
 * @returns An Astro `APIRoute` handler that runs the effect and returns a `Response`.
 */
export const createEffectAPIRoute = (fn: EffectAPIRouteHandler): APIRoute => {
	return async (context: APIContext): Promise<Response> => {
		return await runEffect(fn(context));
	};
};

/**
 * Wraps an API route handler with effectful middleware, including CORS, validation, timeout, and custom hooks.
 *
 * @param fn - The main effectful API route handler function to execute.
 * @param options - Optional configuration for the route, including CORS, validation, timeout, and middleware hooks.
 * @returns An `APIRoute` function compatible with Astro's API routes.
 *
 * @remarks
 * This utility provides a standardized way to handle API routes with common concerns such as:
 * - Applying CORS headers and handling preflight requests.
 * - Running pre-effect middleware (`onBeforeEffect`).
 * - Validating incoming requests (`validate`).
 * - Enforcing a timeout for the effect execution.
 * - Running post-success (`onSuccess`) and error (`onError`) middleware.
 * - Ensuring CORS headers are present on all responses, including errors.
 *
 * @example
 * ```typescript
 * export const POST = withEffect(async (context) => {
 *   // Your effect logic here
 * }, {
 *   cors: { origin: '*' },
 *   validate: { body: z.object({ name: z.string() }) },
 *   timeout: 5000,
 *   onSuccess: async (response, context) => response,
 *   onError: async (error, context) => new Response('Error', { status: 500 }),
 * });
 * ```
 */
export const withEffectAPI = (
	fn: EffectAPIRouteHandler,
	options: EffectRouteOptions = {}
): APIRoute => {
	return async (context: APIContext): Promise<Response> => {
		try {
			// Apply CORS headers if configured
			const corsHeaders = applyCors(context, options.cors);

			// Handle preflight requests
			if (context.request.method === 'OPTIONS') {
				return new Response(null, {
					status: 204,
					headers: corsHeaders,
				});
			}

			// Apply pre-effect middleware
			let processedContext = context;
			if (options.onBeforeEffect) {
				processedContext = await options.onBeforeEffect(context);
			}

			// Validate request if validators provided
			if (options.validate) {
				const validationError = await validateRequest(processedContext, options.validate);
				if (validationError) {
					const errorResponse = new Response(
						JSON.stringify({ error: 'Validation failed', details: validationError }),
						{ status: 400, headers: { 'Content-Type': 'application/json', ...corsHeaders } }
					);
					return errorResponse;
				}
			}

			// Execute the effect with timeout if specified
			let effectPromise = runEffect(fn(processedContext));
			if (options.timeout) {
				effectPromise = Promise.race([
					effectPromise,
					new Promise<never>((_, reject) =>
						setTimeout(() => reject(new Error('Request timeout')), options.timeout)
					),
				]);
			}

			const response = await effectPromise;

			// Add CORS headers to response
			if (corsHeaders && Object.keys(corsHeaders).length > 0) {
				Object.entries(corsHeaders).forEach(([key, value]) => {
					response.headers.set(key, value);
				});
			}

			// Apply success middleware
			return options.onSuccess ? await options.onSuccess(response, processedContext) : response;
		} catch (error) {
			const corsHeaders = applyCors(context, options.cors);

			if (options.onError) {
				const errorResponse = await options.onError(error, context);
				// Ensure CORS headers on error responses too
				Object.entries(corsHeaders).forEach(([key, value]) => {
					errorResponse.headers.set(key, value);
				});
				return errorResponse;
			}

			// Default error handling with CORS
			return new Response(
				JSON.stringify({
					error: error instanceof Error ? error.message : 'Internal Server Error',
				}),
				{
					status: error instanceof Error && error.message === 'Request timeout' ? 408 : 500,
					headers: {
						'Content-Type': 'application/json',
						...corsHeaders,
					},
				}
			);
		}
	};
};

/**
 * Creates API route handlers with effect support for each HTTP method provided.
 *
 * This function takes a set of route handlers and an optional configuration object,
 * then returns an object mapping HTTP methods to their corresponding API route handlers,
 * each wrapped with effect middleware and merged configuration.
 *
 * The return type is precisely typed to only include the methods that were provided
 * in the handlers parameter, eliminating undefined from the union types.
 *
 * @param handlers - An object containing route handlers keyed by HTTP method (e.g., 'GET', 'POST').
 * @param config - Optional global configuration for all handlers. Can include method-specific overrides via the `methods` property.
 * @returns An object mapping each provided HTTP method to its configured API route handler.
 */
export const createEffectAPIRoutes = <T extends RouteHandlers>(
	handlers: T,
	config: RouteHandlerConfig = {}
): ExtractDefinedRoutes<T> => {
	const routes = {} as ExtractDefinedRoutes<T>;

	for (const [method, handler] of Object.entries(handlers)) {
		if (handler) {
			const methodKey = method as keyof RouteHandlers;

			// Merge global config with method-specific overrides
			const methodConfig = {
				...config,
				...(config.methods?.[methodKey] || {}),
			};

			// Remove the methods property from the final config
			const { methods: _droppedMethods, ...finalConfig } = methodConfig;

			// TypeScript knows this is safe because we checked `if (handler)` above
			// biome-ignore lint/suspicious/noExplicitAny: `routes` is typed as `ExtractDefinedRoutes<T>`
			(routes as any)[methodKey] = withEffectAPI(handler as EffectAPIRouteHandler, finalConfig);
		}
	}

	return routes;
};

/**
 * A builder class for defining API route handlers with optional per-method and global configuration.
 *
 * Provides a fluent interface for registering HTTP method handlers (`get`, `post`, `put`, `delete`, `patch`, `all`)
 * and setting global or per-method options. Use `build()` to generate the final route configuration.
 *
 * Each method returns a new builder type that accumulates the added methods, ensuring precise typing
 * in the final built routes object.
 *
 * @example
 * ```typescript
 * const routes = new EffectRouteBuilder()
 *   .withGlobalConfig({ cors: { origin: true } })
 *   .get(getHandler, { timeout: 3000 })
 *   .post(postHandler)
 *   .build();
 * // routes has type: { GET: APIRoute, POST: APIRoute } (no undefined!)
 * ```
 */

// biome-ignore lint/complexity/noBannedTypes: `T` is used to accumulate handlers
export class EffectAPIRouteBuilder<T extends Partial<RouteHandlers> = {}> {
	private handlers: T;
	private config: RouteHandlerConfig = {};

	constructor(handlers: T = {} as T) {
		this.handlers = handlers;
	}

	// Set global configuration
	public withGlobalConfig(config: EffectRouteOptions): EffectAPIRouteBuilder<T> {
		this.config = { ...this.config, ...config };
		return this;
	}

	private addHandler<K extends keyof RouteHandlers, U extends EffectAPIRouteHandler>(
		method: K,
		handler: U,
		options?: Partial<EffectRouteOptions>
	): EffectAPIRouteBuilder<T & Record<K, U>> {
		const newHandlers = { ...this.handlers, [method]: handler } as T & Record<K, U>;
		const newBuilder = new EffectAPIRouteBuilder(newHandlers);
		newBuilder.config = { ...this.config };

		if (options) {
			newBuilder.config.methods = { ...this.config.methods, [method]: options };
		}
		return newBuilder;
	}

	// Add handlers with fluent interface - each method returns a new type that includes the added method
	public get<U extends EffectAPIRouteHandler>(
		handler: U,
		options?: Partial<EffectRouteOptions>
	): EffectAPIRouteBuilder<T & { GET: U }> {
		return this.addHandler('GET', handler, options) as EffectAPIRouteBuilder<T & { GET: U }>;
	}

	public post<U extends EffectAPIRouteHandler>(
		handler: U,
		options?: Partial<EffectRouteOptions>
	): EffectAPIRouteBuilder<T & { POST: U }> {
		return this.addHandler('POST', handler, options) as EffectAPIRouteBuilder<T & { POST: U }>;
	}

	public put<U extends EffectAPIRouteHandler>(
		handler: U,
		options?: Partial<EffectRouteOptions>
	): EffectAPIRouteBuilder<T & { PUT: U }> {
		return this.addHandler('PUT', handler, options) as EffectAPIRouteBuilder<T & { PUT: U }>;
	}

	public delete<U extends EffectAPIRouteHandler>(
		handler: U,
		options?: Partial<EffectRouteOptions>
	): EffectAPIRouteBuilder<T & { DELETE: U }> {
		return this.addHandler('DELETE', handler, options) as EffectAPIRouteBuilder<T & { DELETE: U }>;
	}

	public patch<U extends EffectAPIRouteHandler>(
		handler: U,
		options?: Partial<EffectRouteOptions>
	): EffectAPIRouteBuilder<T & { PATCH: U }> {
		return this.addHandler('PATCH', handler, options) as EffectAPIRouteBuilder<T & { PATCH: U }>;
	}

	public all<U extends EffectAPIRouteHandler>(
		handler: U,
		options?: Partial<EffectRouteOptions>
	): EffectAPIRouteBuilder<T & { ALL: U }> {
		return this.addHandler('ALL', handler, options) as EffectAPIRouteBuilder<T & { ALL: U }>;
	}

	// Build the final routes with precise typing
	public build(): ExtractDefinedRoutes<T> {
		return createEffectAPIRoutes(this.handlers, this.config);
	}
}
