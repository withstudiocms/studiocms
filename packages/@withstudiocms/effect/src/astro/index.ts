export * from './api-route.js';
export * from './context-utils.js';
export * from './middleware.js';
export * from './response-helpers.js';
export * from './types.js';
