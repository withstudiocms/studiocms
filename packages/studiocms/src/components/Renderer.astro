---
import contentRenderer from 'studiocms:renderer/current';

interface Props {
	content: string;
}

const { content } = Astro.props;

// @ts-ignore
const renderedContent = await contentRenderer(content, $$result);
---
<Fragment set:html={renderedContent}></Fragment>