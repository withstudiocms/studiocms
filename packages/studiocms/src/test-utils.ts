import type { AstroIntegration, AstroIntegrationLogger } from 'astro';
import type { StudioCMSPlugin } from './plugins.js';
import type {
	SCMSAuthServiceFnOpts,
	SCMSDashboardFnOpts,
	SCMSFrontendFnOpts,
	SCMSImageServiceFnOpts,
	SCMSRenderingFnOpts,
	SCMSSiteMapFnOpts,
} from './schemas/index.js';

type HookRun<T> = { hasHook: boolean; hookResults: T };
export interface PluginHookResults {
	astroConfig: HookRun<{ integrations: AstroIntegration[] }>;
	studiocmsConfig: HookRun<{
		authService: Partial<SCMSAuthServiceFnOpts>;
		dashboard: Partial<SCMSDashboardFnOpts>;
		frontend: Partial<SCMSFrontendFnOpts>;
		imageService: Partial<SCMSImageServiceFnOpts>;
		rendering: Partial<SCMSRenderingFnOpts>;
		sitemap: Partial<SCMSSiteMapFnOpts>;
	}>;
}
/**
 * Utility class for testing StudioCMS plugins by simulating hook execution and collecting results.
 *
 * The `StudioCMSPluginTester` provides methods to:
 * - Create mock loggers for use in plugin hooks.
 * - Execute the `studiocms:astro:config` and `studiocms:config:setup` hooks, if present, on a given plugin.
 * - Collect and return integration and configuration data set by these hooks.
 * - Retrieve basic plugin metadata.
 *
 * @remarks
 * This class is intended for use in test environments to facilitate inspection and validation of plugin behavior.
 *
 * @example
 * ```typescript
 * const tester = new StudioCMSPluginTester(myPlugin);
 * const info = tester.getPluginInfo();
 * const hookResults = await tester.getHookResults();
 * ```
 */
export class StudioCMSPluginTester {
	private plugin: StudioCMSPlugin;
	private readonly injectedLogger?: AstroIntegrationLogger;

	constructor(plugin: StudioCMSPlugin, logger?: AstroIntegrationLogger) {
		this.plugin = plugin;
		this.injectedLogger = logger;
	}

	/**
	 * Creates a mock logger object for testing purposes.
	 *
	 * The returned logger implements the `AstroIntegrationLogger` interface and provides
	 * stubbed logging methods (`info`, `warn`, `error`, `debug`) that delegate to the corresponding
	 * `console` methods. It also includes a `fork` method, which returns a new mock logger
	 * instance with an attached label for testing logger forking behavior.
	 *
	 * @returns {AstroIntegrationLogger} A mock logger instance suitable for use in tests.
	 */
	private createMockLogger(): AstroIntegrationLogger {
		if (this.injectedLogger) return this.injectedLogger;
		// biome-ignore lint/suspicious/noExplicitAny: this is fine
		const logger: Record<string, any> = {
			info: console.log,
			warn: console.warn,
			error: console.error,
			debug: console.debug,
			fork: (label: string) => {
				// Each fork returns a new mock logger with the label attached for testing
				const forked = this.createMockLogger();
				forked.label = label;
				return forked;
			},
		};
		return logger as unknown as AstroIntegrationLogger;
	}

	/**
	 * Executes the 'studiocms:astro:config' hook if it exists on the plugin,
	 * providing a mock logger and a method to collect Astro integrations.
	 *
	 * @returns A promise that resolves to an object containing the collected integrations.
	 *
	 * @remarks
	 * This method checks if the plugin defines a 'studiocms:astro:config' hook as a function.
	 * If so, it invokes the hook with a mock logger and an `addIntegrations` callback,
	 * which accumulates integrations into an array. The collected integrations are then
	 * returned in an object.
	 */
	private async runAstroConfigHook() {
		const integrations: AstroIntegration[] = [];
		if (
			this.plugin.hooks['studiocms:astro:config'] &&
			typeof this.plugin.hooks['studiocms:astro:config'] === 'function'
		) {
			await this.plugin.hooks['studiocms:astro:config']({
				logger: this.createMockLogger(),
				addIntegrations: (newIntegrations) => {
					const toAdd = Array.isArray(newIntegrations) ? newIntegrations : [newIntegrations];
					integrations.push(...toAdd);
				},
			});
		}
		return { integrations };
	}

	/**
	 * Executes the `studiocms:config:setup` hook on the current plugin, if available,
	 * and collects configuration options set by the hook into partial option objects.
	 *
	 * This method mocks the hook's context by providing setter functions for various
	 * StudioCMS configuration aspects, such as authentication, dashboard, frontend,
	 * image service, rendering, and sitemap. Each setter captures the values provided
	 * by the hook and stores them in corresponding partial option objects.
	 *
	 * @returns A promise that resolves to an object containing the collected configuration
	 *          options for authentication service, dashboard, frontend, image service,
	 *          rendering, and sitemap.
	 *
	 * @remarks
	 * This utility is primarily intended for testing purposes, allowing inspection of
	 * configuration values set by a plugin's `studiocms:config:setup` hook.
	 */
	private async runStudioCMSConfigHook() {
		const authService: Partial<SCMSAuthServiceFnOpts> = {};
		const dashboard: Partial<SCMSDashboardFnOpts> = {};
		const frontend: Partial<SCMSFrontendFnOpts> = {};
		const imageService: Partial<SCMSImageServiceFnOpts> = {};
		const rendering: Partial<SCMSRenderingFnOpts> = {};
		const sitemap: Partial<SCMSSiteMapFnOpts> = {};

		if (
			this.plugin.hooks['studiocms:config:setup'] &&
			typeof this.plugin.hooks['studiocms:config:setup'] === 'function'
		) {
			await this.plugin.hooks['studiocms:config:setup']({
				logger: this.createMockLogger(),
				setAuthService: ({ oAuthProvider }) => {
					if (oAuthProvider) {
						authService.oAuthProvider = oAuthProvider;
					}
				},
				setDashboard: ({ dashboardGridItems, dashboardPages }) => {
					if (dashboardGridItems) {
						dashboard.dashboardGridItems = dashboardGridItems;
					}
					if (dashboardPages) {
						dashboard.dashboardPages = dashboardPages;
					}
				},
				setFrontend: ({ frontendNavigationLinks }) => {
					if (frontendNavigationLinks) {
						frontend.frontendNavigationLinks = frontendNavigationLinks;
					}
				},
				setImageService: ({ imageService: imgService }) => {
					if (imgService) {
						imageService.imageService = imgService;
					}
				},
				setRendering: ({ pageTypes }) => {
					if (pageTypes) {
						rendering.pageTypes = pageTypes;
					}
				},
				setSitemap: ({ sitemaps, triggerSitemap }) => {
					if (sitemaps) {
						sitemap.sitemaps = sitemaps;
					}
					sitemap.triggerSitemap = triggerSitemap;
				},
			});
		}

		return {
			authService,
			dashboard,
			frontend,
			imageService,
			rendering,
			sitemap,
		};
	}

	/**
	 * Retrieves information about the current plugin.
	 *
	 * @returns An object containing the plugin's identifier, name, minimum required StudioCMS version, and dependencies.
	 */
	public getPluginInfo() {
		return {
			identifier: this.plugin.identifier,
			name: this.plugin.name,
			studiocmsMinimumVersion: this.plugin.studiocmsMinimumVersion,
			requires: this.plugin.requires,
		};
	}

	/**
	 * Asynchronously retrieves the results of configured plugin hooks.
	 *
	 * @returns An object containing the presence and results of the 'studiocms:astro:config' and 'studiocms:config:setup' hooks.
	 * - `astroConfig`: Indicates if the 'studiocms:astro:config' hook exists and provides its execution results.
	 * - `studiocmsConfig`: Indicates if the 'studiocms:config:setup' hook exists and provides its execution results.
	 */
	public async getHookResults(): Promise<PluginHookResults> {
		return {
			astroConfig: {
				hasHook: !!this.plugin.hooks['studiocms:astro:config'],
				hookResults: await this.runAstroConfigHook(),
			},
			studiocmsConfig: {
				hasHook: !!this.plugin.hooks['studiocms:config:setup'],
				hookResults: await this.runStudioCMSConfigHook(),
			},
		};
	}
}
