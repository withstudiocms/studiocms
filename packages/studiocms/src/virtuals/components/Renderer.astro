---
import { renderAugments } from 'studiocms:plugins/augments';
import { pluginRenderers } from 'studiocms:plugins/renderers';
import * as renderers from 'virtual:studiocms/plugins/renderers';
import type { SSRResult } from 'astro';
import type { PluginPageTypeRendererProps, PluginRenderer } from '../../types.js';
import renderFn from './renderFn.js';

interface Props extends PluginPageTypeRendererProps {}

// Get props
const { data } = Astro.props as Props;

// If no data, return nothing
if (!data) return;

// Find the renderer for the given page type
const renderer = pluginRenderers.find((r) => r.pageType === data.package);

/* v8 ignore start */
if (!renderer) {
	console.warn(
		`Renderer for page type "${data.package}" not found. Ensure the plugin is correctly registered and the renderer exists.`
	);
	// If no renderer found, return nothing
	return;
}
/* v8 ignore stop */

// Get the renderer module
const RendererModule: PluginRenderer = renderers[renderer.safePageType as keyof typeof renderers];

/* v8 ignore start */
// Check if RendererModule has a name property to verify it's a valid interface
if (!RendererModule.name) {
	console.error(`Renderer module for page type "${data.package}" not found in renderers.`);
	// If no renderer module found, return nothing
	return;
}
/* v8 ignore stop */

// Filter augments to only include those specified in the page data
const augments = renderAugments.filter((augment) =>
	(data.augments as string[])?.includes(augment.id)
);

// Render the content using the render function
// If no content found, return an error message
// Pass in the preRenderer and sanitizeOpts if they exist on the RendererModule
const renderedContent = await renderFn({
	content: data.defaultContent?.content || '<h1>Error: No content found</h1>',
	augments,
	renderOpts: {
		// @ts-expect-error - $$result is a global variable injected by Astro during compilation and not a editor-known variable
		result: $$result as SSRResult,
		sanitizeOpts: RendererModule.sanitizeOpts,
		preRenderer: RendererModule.renderer,
	},
});
---

<Fragment set:html={renderedContent} />
