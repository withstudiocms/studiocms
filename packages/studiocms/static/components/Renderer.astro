---
/// <reference types="../../renderer.d.ts" />
import contentRenderer from 'studiocms:renderer/current';

interface Props {
	content: string;
}

const { content } = Astro.props;

// @ts-expect-error - `$$result` is a variable injected by the Astro compiler
const renderedContent = await contentRenderer(content, $$result);
---
<Fragment set:html={renderedContent} />