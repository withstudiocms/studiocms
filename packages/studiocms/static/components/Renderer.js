import contentRenderer from 'studiocms:renderer/current';
import { HTMLString } from 'astro/runtime/server/index.js';

export default Object.assign(
	function Renderer(result, { content }) {
		return {
			get [Symbol.toStringTag]() {
				return 'AstroComponent';
			},
			async *[Symbol.asyncIterator]() {
				const renderedContent = await contentRenderer(content, result);
				yield new HTMLString(renderedContent);
			},
		};
	},
	{
		isAstroComponentFactory: true,
	}
);
