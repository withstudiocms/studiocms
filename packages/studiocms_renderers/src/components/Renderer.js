import { logger } from '@it-astro:logger:studiocms-renderer';
import rendererConfig from 'studiocms:renderer/config';
import { HTMLString } from 'astro/runtime/server/index.js';
import builtInContentRenderer, { contentRenderer } from '../lib/contentRenderer';

globalThis.StudioCMS_Renderer = contentRenderer;

export const Renderer = Object.assign(
	function Renderer(result, attributes, slots) {
		return {
			get [Symbol.toStringTag]() {
				return 'AstroComponent';
			},
			async *[Symbol.asyncIterator]() {
				const content = attributes.content;
				yield new HTMLString(await globalThis.StudioCMS_Renderer(content));
			},
		};
	},
	{
		isAstroComponentFactory: true,
	}
);

export default Renderer;
