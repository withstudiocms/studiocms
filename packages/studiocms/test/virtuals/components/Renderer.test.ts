/// <reference types="astro/client" />
import { experimental_AstroContainer as AstroContainer } from 'astro/container';
import { describe, expect, test } from 'vitest';
import Renderer from '../../../src/virtuals/components/Renderer.astro';
import { cleanAstroAttributes, makeRendererProps } from '../../test-utils';

describe('Renderer component', () => {
	test('render component with content', async () => {
		const container = await AstroContainer.create();
		const result = await container.renderToString(Renderer, {
			props: makeRendererProps('<div>This is a test content from the defaultContent field.</div>'),
		});
		const cleanResult = cleanAstroAttributes(result, '/mock/path/Renderer.astro');
		expect(cleanResult).toMatchSnapshot();
	});

	test('render component with no content', async () => {
		const container = await AstroContainer.create();
		const result = await container.renderToString(Renderer, {
			props: makeRendererProps(null),
		});
		const cleanResult = cleanAstroAttributes(result, '/mock/path/Renderer.astro');
		expect(cleanResult).toMatchSnapshot();
	});
});
