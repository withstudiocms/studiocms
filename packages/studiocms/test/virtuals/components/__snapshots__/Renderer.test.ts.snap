// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Renderer component > render component 1`] = `"<div>This is a test content from the defaultContent field.</div>"`;

exports[`Renderer component > render component with no content 1`] = `"<h1>Error: No content found</h1>"`;
